File: src/day9.ts
const solution = {
    part1: 0,
    part2: 0,
};

const text = await Bun.file("input/day9.txt").text();

function extrapolate(nums: number[], first: boolean): number {
    let all_zero = true;
    const deltas = [];

    for (let i = 1; i < nums.length; ++i) {
        deltas.push(nums[i] - nums[i - 1]);
        if (deltas.at(-1) !== 0)
            all_zero = false;
    }

    if (all_zero)
        return first ? nums.at(0)! : nums.at(-1)!;

    if (first)
        return nums.at(0)! - extrapolate(deltas, first);

    return nums.at(-1)! + extrapolate(deltas, first);
}

for (const l of text.trim().split("\n")) {
    const nums = l.split(" ").map((s) => parseInt(s));
    solution.part1 += extrapolate(nums, false);
    solution.part2 += extrapolate(nums, true);
}

console.log(solution);

File: aoc2016/src/day9.rs
enum Version {
    V1,
    V2,
}

pub fn part1(input: &str) -> usize {
    decompressed_len(input, Version::V1)
}

pub fn part2(input: &str) -> usize {
    decompressed_len(input, Version::V2)
}

fn decompressed_len(input: &str, version: Version) -> usize {
    let mut total_len = 0;

    let mut chars = input.trim().chars();
    let chars = chars.by_ref();
    while let Some(ch) = chars.next() {
        if ch != '(' {
            total_len += 1;
            continue;
        }

        let marker = chars.take_while(|c| *c != ')').collect::<String>();

        let mut parts = marker.split('x');
        let len: usize = parts.next().unwrap().parse().unwrap();
        let repeat: usize = parts.next().unwrap().parse().unwrap();

        let sub_len = match version {
            Version::V1 => {
                chars.take(len).last();
                len
            }
            Version::V2 => {
                let sub = chars.take(len).collect::<String>();
                decompressed_len(&sub, Version::V2)
            }
        };

        total_len += sub_len * repeat;
    }

    total_len
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part1() {
        assert_eq!(part1(include_str!("../input/day9.txt")), 70186);
    }

    #[test]
    fn test_part2() {
        assert_eq!(part2(include_str!("../input/day9.txt")), 10915059201);
    }
}
